core/src/smartcontracts/isi/mod.rs

```diff
@@ -267,7 +266,7 @@ impl Execute for SetKeyValueBox {
             IdBox::DomainId(id) => {
                 SetKeyValue::<Domain, Name, Value>::new(id, key, value).execute(authority, wsv)
             }
-            _ => Err(Error::Unsupported(InstructionType::SetKeyValue)),
+            _ => Err(Error::Evaluate(InstructionType::SetKeyValue.into())),
```

Comment:
Why so? I think `Unsupported` error clearly shows that this instructions with this parameter types is not supported. I don't quite understand what `Evaluate` error means here. Maybe it least something like `Evaluate(Unsupported(...))`?

Reply:
IMO `Unsupported` is a type of evaluation error because your expression failed to produce a valid type. `Evaluate` is about whether the instruction is valid in the sense that there is at least one state of Iroha that it can be executed against. If evaluation fails this means that instruction is not a valid construct according to the specification of Iroha DSL. 

`Validate` on the other hand is a separate step that verifies that a valid instruction can be executed at this time on the current state of Iroha. This is about permissions or invalid operations like divide amount in metadata by 0, etc.

Interestingly, `Expression::Query` is the only expression that requires `wsv`. If we wanted to pre-evaluate instructions we could do so by reducing every expression either to a value or to a `f(query)`. This would bring us a benefit of smaller chain size and less repeated calculation. In this case we can even consider client side evaluation